pipelines/common/src/main/java/org/openmrs/analytics/DwhFiles.java

```diff
@@ -115,32 +113,58 @@ public Path getIncrementalRunPathWithTimestamp() {
     if (hasIncrementalDir()) {
-      Path movePath = getIncrementalRunPathWithTimestamp();
+      ResourceId movePath = getIncrementalRunPathWithTimestamp();
       log.info("Moving the old {} directory to {}", INCREMENTAL_DIR, movePath);
-      Files.move(incPath, movePath);
+      FileSystems.rename(Collections.singletonList(incPath), Collections.singletonList(movePath));
     }
     return incPath;
   }
 
   /** @return true iff there is already an incremental run subdirectory in this DWH. */
-  public boolean hasIncrementalDir() {
-    return Files.exists(getIncrementalRunPath());
+  public boolean hasIncrementalDir() throws IOException {
+    List<MatchResult> matches =
+        FileSystems.matchResources(Collections.singletonList(getIncrementalRunPath()));
+    MatchResult matchResult = Iterables.getOnlyElement(matches);
+    if (matchResult.status() == Status.OK) {
+      return true;
+    }
+    return false;
   }
 
-  public Set<String> findResourceTypes() throws IOException {
-    // TODO: Switch to using Beam's FileSystems API or make sure this works on all platforms.
-    //   Currently Beam's API fails to find directories under `dwhPath` with "*" wildcard and
-    //   even with ResourceId.resolve; why? Sample pseudo-code:
-    //   MatchResult matchResult = FileSystems.match(dwhRoot); // with "/*" it does not work either!
-    //   ResourceId resourceId = matchResult -> metadata -> resourceId;
-    //   resourceId.resolve("*", StandardResolveOptions.RESOLVE_DIRECTORY)
-    //   https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/issues/390
-    Path dwhPath = fileSystem.getPath(dwhRoot);
-    Set<String> fileSet =
-        Files.list(dwhPath).map(p -> dwhPath.relativize(p).toString()).collect(Collectors.toSet());
+  /**
+   * This method returns the list of non-empty directories which contains atleast one file under it.
+   * The directory name should be a valid FHIR resource type.
+   *
+   * @return the set of non-empty directories
+   * @throws IOException
+   */
+  public Set<String> findNonEmptyFhirResourceTypes() throws IOException {
+    // TODO : If the list of files under the dwhRoot is huge then there can be a lag in the api
+    //  response. This issue https://github.com/google/fhir-data-pipes/issues/288 helps in
+    //  maintaining the number of file to an optimum value.
+
+    ResourceId childResourceId =
+        FileSystems.matchNewResource(dwhRoot, true)
+            .resolve("*/*", StandardResolveOptions.RESOLVE_FILE);
+
+    Set<String> fileSet = new HashSet<>();
+    List<MatchResult> matchedChildResultList =
+        FileSystems.matchResources(Collections.singletonList(childResourceId));
+
+    for (MatchResult matchResult : matchedChildResultList) {
+      if (matchResult.status() == Status.OK && !matchResult.metadata().isEmpty()) {
+        for (Metadata metadata : matchResult.metadata()) {
+          fileSet.add(metadata.resourceId().getCurrentDirectory().getFilename());
```

Comment:
Question: My understanding was that GCS has a flat namespace, i.e., there is no real "directory" per se. But from this code here, it seems it simulates the notion of directory, right? I mean, `getCurrentDirectory` for `my_dir/Patient` will return `my_dir`, correct?

Reply:
Yes, your understanding is correct. GCS has flat namespace and only simulates the notion of directory. Here, `getCurrentDirectory` for `my_dir/Patient` will return `my_dir`